Feat: fallback to CLI config for token

Praveen005 opened the pull request:
This pull request introduces the feature requested in #246 

It Implements a priority-based token retrieval:
1. Environment variable (CIVO_TOKEN).
2. Token provided via a Credential file.
3. Lastly, fall back to CLI config file (~/.civo.json)

**Tests'  Screenshots:**

I have introduced coloured debug messages(removed from the main code), which correctly prints the token it abstracts:

<img width="602" alt="Screenshot 2024-07-08 213657" src="https://github.com/civo/terraform-provider-civo/assets/42026111/6e9775bb-ad6f-4894-b036-ee47e33ebd6b">

Below, I have set the token using Civo CLI command `civo apikey save CIVO_TOKEN clitoken96` which saves it in `~/.civo.json`.
You can see, it correctly prints a different token(last coloured debug message):

<img width="616" alt="Screenshot 2024-07-08 214536" src="https://github.com/civo/terraform-provider-civo/assets/42026111/59990e4f-82d9-4194-b7ab-30f84efed621">

uzaxirr commented:
@Praveen005 is this ready for review?

Praveen005 commented:
Yes @uzaxirr it is ready for review.

uzaxirr commented:
@Praveen005  can you also build the provider locally, test it and attach corresponding screenshots.
Refer to [this](https://github.com/civo/terraform-provider-civo?tab=readme-ov-file#local-development-using-a-local-version-of-the-provider) for local setup.

uzaxirr commented on civo/provider.go:
```diff
@@ -34,11 +37,11 @@ var (
 	return &schema.Provider{
 		Schema: map[string]*schema.Schema{
-			"token": {
+			"credential_file": {
 				Type:        schema.TypeString,
 				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("CIVO_TOKEN", ""),
-				Description: "This is the Civo API token. Alternatively, this can also be specified using `CIVO_TOKEN` environment variable.",
```
Instead of removing token, we can deprecate this field from the schema. This would be a better way.

```
			"token": {
				Type:        schema.TypeString,
				Optional:    true,
				DefaultFunc: schema.EnvDefaultFunc("CIVO_TOKEN", ""),
				Deprecated: "\"token\" attribute is deprecated. Moving forward, please use \"credential_file\" attribute.",
				Description: "This is the Civo API token. Alternatively, this can also be specified using `CIVO_TOKEN` environment variable.",
			},
```

Praveen005 replied:
Hey @uzaxirr I zeroed in on the likely bottleneck,

The `token` variable in `provider.tf` in `build.sh` script was still there, I removed it.

Also updated the `provider.tf` in `examples` directory.

I also incorporated your suggestion of deprecating the token field. Thanks for the heads up.

Following are the screen grabs from local build:
<img width="718" alt="Screenshot 2024-07-09 172827" src="https://github.com/civo/terraform-provider-civo/assets/42026111/0ced17f1-49f8-48c5-9348-f2747f28b8b5">
<img width="833" alt="Screenshot 2024-07-09 172959" src="https://github.com/civo/terraform-provider-civo/assets/42026111/b851476f-acab-4066-851d-260a740f5d54">
<img width="811" alt="Screenshot 2024-07-09 173037" src="https://github.com/civo/terraform-provider-civo/assets/42026111/aed9d5b1-36fb-400f-aa87-454418a5a9d6">


The resource didn't get created, likely because I don't have an active Civo Account.
Can you now try again?

Regards.

uzaxirr replied:
While you performed the above operation was there any `token` or credential_file attribute present in your provider.tf civo config block?

Praveen005 replied:
@uzaxirr No, it was not in `provider.tf`. I will share you the terraform debug log which shows token being read from `~/.civo.json`. The code need some more changes and I am working on that.

uzaxirr replied:
Sure, great work so far

uzaxirr reviewed (changes requested):
I've tested your branch locally. It works when `credential_file` is specified or env vars are there but it fails when it has to fallback to CLI configs i.e when credential_file is not specified and env vars are also absent. can you please verify this.

Praveen005 commented:
Hi @uzaxirr Thank you for your patience.

Here are the details about the changes:

1. Earlier commit was not reading token from the `token` attribute, so fixed it.
2. Added one test case, to cater for this change.
3. In last commit I removed, `civo_api_key` and `token = "$civo_api_key"` from the local build script `local.sh` but that's not needed, so added them back.
4. The code now, passes the local build test, all the unit tests and also correctly reads the token from `~/.civo.json`

**Screenshots:**

1. `provider.tf` file on which I'm running the local build:
You can see, I haven't provided it the `token` nor the `credential_file` path. I haven't set the environment variable either.
Meaning it will have to fall back to reading from `~/.civo.json`

<img width="752" alt="Screenshot 2024-07-10 130720" src="https://github.com/civo/terraform-provider-civo/assets/42026111/8090b970-4e14-46bf-bf3e-39c89b189fa9">

2. No environment variable set:

<img width="627" alt="Screenshot 2024-07-10 131059" src="https://github.com/civo/terraform-provider-civo/assets/42026111/ad2b1faf-69d3-41f6-8c9a-c6b938887952">

3. Token in `~/.civo.json`

<img width="638" alt="Screenshot 2024-07-10 131319" src="https://github.com/civo/terraform-provider-civo/assets/42026111/95fa5671-2da5-4131-847e-85ee4f0b43a4">

4. Added debug log(removed from main code) to print the token when it is read:
<img width="721" alt="Screenshot 2024-07-10 123037" src="https://github.com/civo/terraform-provider-civo/assets/42026111/e3e721ec-060e-40b5-b671-71bed1949642">

5. Starting local build:
<img width="649" alt="Screenshot 2024-07-10 123512" src="https://github.com/civo/terraform-provider-civo/assets/42026111/eac91db8-bbb2-4a05-b145-319050eb6f29">

6. Terraform debug log, shows `CIVO_TOKEN` being correctly read from CLI config.
<img width="903" alt="Screenshot 2024-07-10 123606" src="https://github.com/civo/terraform-provider-civo/assets/42026111/cdba733e-5d54-46bc-bb4e-b80689378b28">


7. See deprecation warning being correctly displayed:
<img width="776" alt="Screenshot 2024-07-10 123628" src="https://github.com/civo/terraform-provider-civo/assets/42026111/cf661f01-2895-4eef-bc75-9ba9965710e2">

8. Terraform apply:
<img width="597" alt="Screenshot 2024-07-10 123719" src="https://github.com/civo/terraform-provider-civo/assets/42026111/84c1a245-fb63-4032-8bfc-e2444594cadd">

9. Token again being read again after `terraform apply` command:
<img width="901" alt="Screenshot 2024-07-10 123740" src="https://github.com/civo/terraform-provider-civo/assets/42026111/a3c15739-fefc-4a1e-91fa-92ece95d95eb">

10. Attempts being made to create the resource, but fails because I don't have an active civo account
<img width="905" alt="Screenshot 2024-07-10 123828" src="https://github.com/civo/terraform-provider-civo/assets/42026111/51d6a589-19e5-4f75-bbf1-501723505857">

11. Passing all the unit tests:
<img width="666" alt="Screenshot 2024-07-10 124039" src="https://github.com/civo/terraform-provider-civo/assets/42026111/697fc6d2-5abd-48f9-b95c-1b6164eeff7c">

12. Test coverage:
<img width="645" alt="Screenshot 2024-07-10 124059" src="https://github.com/civo/terraform-provider-civo/assets/42026111/2940858d-236c-4239-8a6b-fca344037fe8">


Can you please review these changes again?

Regards.

uzaxirr reviewed (changes requested):
Hey @Praveen005 
I've pulled up your latest changes locally, again the same scenario works fine when `crediential_file` or `CIVO_TOKEN` env var is specified. but fails to obtain the token from CLI when nothing is specified. Please refer to the below screen shot

![Screenshot 2024-07-15 at 10 43 56 AM](https://github.com/user-attachments/assets/3221621f-1e1e-4871-846d-503166d45c9c)

Also why do i see the `Argument is deprecated` warning even though i haven't specified any token field in my schema.

I would suggest you to create a CIVO account, set up the CLI and then try spinning up resources via terraform with all the different scnerios.

Praveen005 commented:
I will check it right away, @uzaxirr.

Praveen005 commented:
Hi @uzaxirr, sorry it took way more time than I anticipated.

I believe, it fixes all the above issues you pointed. I have tested it thoroughly and it works fine.

1. I added a validation function under the `token` attribute in provider schema. The warnings are now logged only when you use the `token` variable.
If you see the `token` field, I removed the deprecated message, because the warning was shown twice in quick succession, once because I used `Deprecated` attribute, while other in `validateTokenUsage` function.

2. Earlier I had hardcoded the API key name to `CIVO_TOKEN`, and this was causing the trouble while reading the token from cli config. on your machine. I corrected it.

3. This change now will enable the user to use multiple API keys.

4. The credential file now will look like:
```
{
				"apikeys": {
					"my_token1": "123456",
					"my_token2": "785645"
				},
				"meta": {
					"current_apikey": "my_token1"
				}
}
```

Screenshots:

`provider.tf`

<img width="747" alt="Screenshot 2024-07-15 222200" src="https://github.com/user-attachments/assets/5a3172de-6a01-4e61-a374-470de71cb197">

Network resource successfully created:

<img width="760" alt="Screenshot 2024-07-15 222449" src="https://github.com/user-attachments/assets/c5680286-4846-4ce1-ac81-d23922c81e80">

uzaxirr commented:
@Praveen005 can you test the following locally and attach screenshots for them.

1. with credential_file attribute
2. with token attribute
3. with just the CIVO_TOKEN env var, no credential file or token
4. Without credential_file, token or the env var (i.e token should be picked up from the CLI config)
5. without  credential_file, token, env var or ~/.civo.json file present (i.e some error should occour)

Praveen005 commented:
Thank you for reviewing it @uzaxirr.

Here are all the relevant screenshots:

1. Token provided only via `credential_file` attribute.
<img width="795" alt="Screenshot 2024-07-17 023709" src="https://github.com/user-attachments/assets/e5894d39-df76-4dd6-afdd-c3ff064e9e0e">
<img width="584" alt="Screenshot 2024-07-17 023821" src="https://github.com/user-attachments/assets/2c57b821-9e5b-419e-b291-4deb66f28ddf">

2. Token provided only via `token` attribute
<img width="790" alt="Screenshot 2024-07-17 024222" src="https://github.com/user-attachments/assets/6f9c15f8-af99-4aa1-8b47-9384acf978dd">
Warning message being correctly displayed:
<img width="737" alt="Screenshot 2024-07-17 024312" src="https://github.com/user-attachments/assets/b23b0c69-729a-4f8d-8c86-c4422130f18f">
<img width="726" alt="Screenshot 2024-07-17 024408" src="https://github.com/user-attachments/assets/f9c632fb-c90e-4606-a6b4-70dca32511f5">



3. Token provided only via `CIVO_TOKEN` environment variable
<img width="924" alt="Screenshot 2024-07-17 023227" src="https://github.com/user-attachments/assets/d5f080d9-be9e-43e6-904d-cf858e54ef6d">
<img width="581" alt="Screenshot 2024-07-17 023252" src="https://github.com/user-attachments/assets/a11445a2-212c-4595-97f6-f6141f7ef679">


4. Token available only in `~/.civo.json`
<img width="786" alt="Screenshot 2024-07-17 005638" src="https://github.com/user-attachments/assets/f98513d6-0800-479e-a13a-9e9d64ec9b1a">
<img width="789" alt="Screenshot 2024-07-17 021108" src="https://github.com/user-attachments/assets/109a8af5-2a4a-420f-b1ab-76ab705058fc">

5. No token Provider at all.
<img width="792" alt="Screenshot 2024-07-17 022426" src="https://github.com/user-attachments/assets/1b712228-9c78-4a43-9210-e63aacce910e">

Resource creation failed:

<img width="539" alt="Screenshot 2024-07-17 022554" src="https://github.com/user-attachments/assets/9262dd0d-86f8-45ae-8b28-0fc5149db90d">

fernando-villalba commented:
Looks great to me so far. As a suggested improvement, if no token is found at all. We could point the user to where to get one:

https://dashboard.civo.com/security

For example:

"No token configuration found in $CIVO_TOKEN  or ~/.civo.json, please go to https://dashboard.civo.com/security to fetch one"

Praveen005 commented:
Hi @fernando-villalba,

I’ve updated the error message to be more expressive as suggested. Here is the new version:

<img width="780" alt="Screenshot 2024-07-17 142936" src="https://github.com/user-attachments/assets/1f0748c1-ea02-4ed7-b938-66cef123c39f">

Can you kindly, check it now?

Thank you!

fernando-villalba commented:
LGTM!

Praveen005 commented:
Hi @alessandroargentieri,

I hope you're doing well. When you have a moment, could you please review this PR?

Thank you!

uzaxirr commented:
hey @Praveen005 the PR looks fine, i'm in process of updating the corresponding docs for this before we merge off this one

Praveen005 commented:
Thank you @uzaxirr. Let me know if there's anything I can assist with regarding the documentation.

Regards!

fernando-villalba commented:
One more thing, when we have the updated documents, I would also add a link to the documentation in the deprecation message. Thank you for your hard work and patience @Praveen005 !

Praveen005 commented:
@fernando-villalba That would be better.

Just a caveat, the link to the documentation will have to be added in `validateTokenUsage` function. 


Also, in the documentation, should we mention the format in which user should store their token in the credential file?
Like:
```
{
  "apikeys": {
    "demo_key": "2p35nGkS9NQncivoko5qqX3FtF643RUfqjniiH2"
  },
  "meta": {
    "current_apikey": "demo_key"
  }
}
```

This way user can store multiple API keys for multiple accounts (home, work etc.). Also it would be consistent with the way, Civo CLI stores token in `~/.civo.json`

fernando-villalba commented:
By the way, @Praveen005 and @uzaxirr 

Sorry to be a stickler for detail but change `credential_file` to `credentials_file` as originally proposed on the ticket.

I know that technically we are only using one credential, the one token, but in a security context is weird referring to it as singular. 

Other than that, let's get this out on next release - doing documentation for it now!

Praveen005 commented:
Hey @uzaxirr, I have tested it again and it works as intended.

with `credentials_file` attribute:

<img width="348" alt="Screenshot 2024-08-07 194640" src="https://github.com/user-attachments/assets/24fdf726-cc5d-4150-be85-68022ae81c12">

with `credential_file` attribute:

<img width="677" alt="Screenshot 2024-08-07 194725" src="https://github.com/user-attachments/assets/15e4543e-e728-4f37-8fda-5d477eabc9ad">

uzaxirr reviewed (approved):
Thankss a ton for this PR

Praveen005 commented:
Thank you @uzaxirr . It means a lot.
